modules/apps/static/portal/portal-db-partition-impl/src/main/java/com/liferay/portal/db/partition/internal/DBPartitionHelperImpl.java

```diff
@@ -0,0 +1,212 @@
+import com.liferay.portal.kernel.model.CompanyConstants;
+import com.liferay.portal.kernel.security.auth.CompanyThreadLocal;
+import com.liferay.portal.kernel.util.GetterUtil;
+import com.liferay.portal.kernel.util.InfrastructureUtil;
+import com.liferay.portal.kernel.util.PropsUtil;
+
+import java.sql.Connection;
+import java.sql.DatabaseMetaData;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.sql.Statement;
+
+import org.osgi.framework.BundleContext;
+import org.osgi.framework.ServiceRegistration;
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Deactivate;
+
+/**
+ * @author Alberto Chaparro
+ */
+@Component(immediate = true, service = {})
+public class DBPartitionHelperImpl implements DBPartitionHelper {
+
+	@Override
+	public boolean addPartition(long companyId) {
+		if (companyId == _defaultCompanyId) {
+			return false;
+		}
+
+		Connection connection = CurrentConnectionUtil.getConnection(
+			InfrastructureUtil.getDataSource());
+
+		String schemaName = "company" + companyId;
+
+		try {
+			try (PreparedStatement preparedStatement =
+					connection.prepareStatement(
+						"create schema if not exists " + schemaName +
+							" character set utf8")) {
+
+				preparedStatement.executeUpdate();
+			}
+
+			DatabaseMetaData databaseMetaData = connection.getMetaData();
+
+			DBInspector dbInspector = new DBInspector(connection);
+
+			try (ResultSet resultSet = databaseMetaData.getTables(
+					dbInspector.getCatalog(), dbInspector.getSchema(), null,
+					new String[] {"TABLE"});
+				Statement statement = connection.createStatement()) {
+
+				while (resultSet.next()) {
+					String tableName = resultSet.getString("TABLE_NAME");
+
+					if (_isControlTable(dbInspector, tableName)) {
+						statement.executeUpdate(
+							StringBundler.concat(
+								"create view ", schemaName, StringPool.PERIOD,
+								tableName, " as select * from companyDefault.",
+								tableName));
+					}
+					else {
+						statement.executeUpdate(
+							StringBundler.concat(
+								"create table ", schemaName, StringPool.PERIOD,
+								tableName, " like companyDefault.", tableName));
+					}
+				}
+			}
+		}
+		catch (Exception exception) {
+			throw new ORMException(exception);
+		}
+
+		return true;
+	}
+
+	@Override
+	public boolean removePartition(long companyId) {
+		return true;
+	}
+
+	@Override
+	public void setDefaultCompanyId(long companyId) {
+		_defaultCompanyId = companyId;
+	}
+
+	@Override
+	public void usePartition(Connection connection) {
+		try {
+			if (connection.isReadOnly()) {
+				return;
+			}
+
+			long companyId = CompanyThreadLocal.getCompanyId();
+
+			try (Statement statement = connection.createStatement()) {
+				if ((companyId == CompanyConstants.SYSTEM) ||
+					(companyId == _defaultCompanyId)) {
+
+					statement.execute("USE companyDefault");
```

Comment:
@achaparro to use this feature, what will you put in the jdbc url? Won't you still need to point it to a schema from the jdbc url, if so, shouldn't it be this "companyDefault" one?

Reply:
Yes, the jdbc url will put CompanyDefault. I always set the the database in the connection (with the use statement) to be sure that we point to the proper database anytime, even when we reuse a connection, do you think we don't need to set it since we always close the connection after its usage?

Reply:
If we could do that the logic would simplify and we could avoid a lot of calls to send the use database command

Reply:
yeah, we should avoid doing this. If this turns out to be needed, that is just a big sign of there is a bug in somewhere.

Reply:
And I think the real problem here is, if you hardcode this "companyDefault" here, does it mean the user must use this schema name? what if they want to use a different name?

Reply:
Ok, I will make this change